pkg/distributor/query.go

```diff
@@ -0,0 +1,187 @@
+			return promql.ErrStorage(err)
+		}
+		return nil
+	})
+	return result, err
+}
+
+func (d *Distributor) queryPrep(ctx context.Context, from, to model.Time, matchers ...*labels.Matcher) (ring.ReplicationSet, *client.QueryRequest, error) {
+	var replicationSet ring.ReplicationSet
+	userID, err := user.ExtractOrgID(ctx)
+	if err != nil {
+		return replicationSet, nil, err
+	}
+
+	req, err := ingester_client.ToQueryRequest(from, to, matchers)
+	if err != nil {
+		return replicationSet, nil, err
+	}
+
+	// Get ingesters by metricName if one exists, otherwise get all ingesters
+	metricNameMatcher, _, ok := extract.MetricNameMatcherFromMatchers(matchers)
+	if !d.cfg.ShardByAllLabels && ok && metricNameMatcher.Type == labels.MatchEqual {
+		replicationSet, err = d.ring.Get(shardByMetricName(userID, []byte(metricNameMatcher.Value)), ring.Read)
+	} else {
+		replicationSet, err = d.ring.GetAll()
+	}
+	return replicationSet, req, err
+}
+
+// queryIngesters queries the ingesters via the older, sample-based API.
+func (d *Distributor) queryIngesters(ctx context.Context, replicationSet ring.ReplicationSet, req *client.QueryRequest) (model.Matrix, error) {
+	// Fetch samples from multiple ingesters in parallel, using the replicationSet
+	// to deal with consistency.
+	results, err := replicationSet.Do(ctx, func(ing *ring.IngesterDesc) (interface{}, error) {
+		client, err := d.ingesterPool.GetClientFor(ing.Addr)
+		if err != nil {
+			return nil, err
+		}
+
+		resp, err := client.(ingester_client.IngesterClient).Query(ctx, req)
+		ingesterQueries.WithLabelValues(ing.Addr).Inc()
+		if err != nil {
+			ingesterQueryFailures.WithLabelValues(ing.Addr).Inc()
+			return nil, err
+		}
+
+		return ingester_client.FromQueryResponse(resp), nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Merge the results into a single matrix.
+	fpToSampleStream := map[model.Fingerprint]*model.SampleStream{}
+	for _, result := range results {
+		for _, ss := range result.(model.Matrix) {
+			fp := ss.Metric.Fingerprint()
+			mss, ok := fpToSampleStream[fp]
+			if !ok {
+				mss = &model.SampleStream{
+					Metric: ss.Metric,
+				}
+				fpToSampleStream[fp] = mss
+			}
+			mss.Values = util.MergeSampleSets(mss.Values, ss.Values)
+		}
+	}
+	result := model.Matrix{}
+	for _, ss := range fpToSampleStream {
+		result = append(result, ss)
+	}
+
+	return result, nil
+}
+
+// queryIngesterStream queries the ingesters using the new streaming API.
+func (d *Distributor) queryIngesterStream(ctx context.Context, replicationSet ring.ReplicationSet, req *client.QueryRequest) ([]client.TimeSeriesChunk, error) {
+	// Fetch samples from multiple ingesters
+	results, err := replicationSet.Do(ctx, func(ing *ring.IngesterDesc) (interface{}, error) {
+		client, err := d.ingesterPool.GetClientFor(ing.Addr)
+		if err != nil {
+			return nil, err
+		}
+
+		stream, err := client.(ingester_client.IngesterClient).QueryStream(ctx, req)
+		ingesterQueries.WithLabelValues(ing.Addr).Inc()
```

Comment:
super nit: Move this one line above? It'll keep the err and it's handling near each other.